### ISSUE TITLE
get_st_ed and ....

### ISSUE THREAD

**Issue by MazenHamdi** [Majiker/STAR#13]
Hello, thank you for sharing the code, this is the best code that I can find on this topic.
I have a couple of questions, and I will appreciate it if you can answer them:
1/ 
![image](https://user-images.githubusercontent.com/37674531/123890176-1608f500-d957-11eb-9894-c4460a400f42.png)
what does this function returns exactly? and what is the use of it ? size of the list is different from the number of frames (20) that we are working on.

2/
![image](https://user-images.githubusercontent.com/37674531/123890310-49e41a80-d957-11eb-9694-dfa9e3a07322.png)
is there a specific reason for initializing the temporal and spatial transformers differently ? is there a difference between the 2 transformers (except the inputs) ?

3/ for the temporal encoder 1, the shape of the input changes in every iteration : (1,N,2) , (2,N,2) ......... (19,N,2)
why would you progressively add a frame each iteration while you are using a GM that stacks any new information each iteration ?
4/Why do you calculate the neighbors number (nei_num) but it was not used ?

5/I am a bit confused about the terms used, what is the difference between frame, scene, window exactly.

Thank you very much, I know it's going to take a lot of your time but I would appreciate it a lot.

**Comment by cunjunyu** [comment]
1. During the training phase, one batch is composed of people from different scenes. This function is used to identify people in the batch from different scenes to do the coordinate normalization. 

2. The forward function is different. In the spatial transformer, we run the encoder with a mask in order to indicate the neighbors of the pedestrian. 

3. I didn't fully get the question so the answer may not be precise. In GM, we only have information prior to the current frame, therefore, we have to stack the information of the current frame to the GM.

4. It is for debugging purposes. 

5. Frame: a timestamp 
    Window: a series of consecutive frames of a certain length 
    Scene: a place during a certain time period (a window)

**Comment by MazenHamdi** [comment]
Thank you very much, that certainly cleared most of the ambiguities,
However, there is one thing that I still do not fully get, the nodes in the batch have the shape (20,N,2), which is logic because the sequence length in our case is 20, (8 frames for observation and 12 frames for prediction, but the batch pedestrian number (batch_ped_number) is not a tensor of shape (20), 
![image](https://user-images.githubusercontent.com/37674531/124213548-10dfad80-daf1-11eb-9a45-7932d819fd4b.png)
isn't the batch_ped_num supposed to contain the number of pedestrians in each frame of the 20 frames ? 
And when a sum of pedestrians in the same frame are considered in the same scene ?

**Comment by cunjunyu** [comment]
> isn't the batch_ped_num supposed to contain the number of pedestrians in each frame of the 20 frames?

Yes, it does. This is recalculated at each iteration.

**Comment by MazenHamdi** [comment]
yes, I agree, but as you can see in the picture of the batch_ped_num, it contains more than 20 numbers, it has 27 elements and not 20 elements (one for each frame)

**Comment by cunjunyu** [comment]
27 elements mean that pedestrians from 27 scenes are used to form this batch. Batch_ped_num is not calculated at once for all 20 frames. It will be updated at each iteration.

**Comment by MazenHamdi** [comment]
what is the difference between saying "pedestrians from 27 scenes are used to form this batch" and "pedestrians from 20 frames are used to form the batch"

**Comment by cunjunyu** [comment]
> Scene: a place during a certain time period (a window)

Each scene has 20 frames.

**Comment by MazenHamdi** [comment]
Thank you very much, I really appreciate your comments and the time invested for this issue, so to sum it up, and correct me if I am wrong please:

we said earlier, in our case for example, 27 different scenes form our batch, and we said that each scene has 20 frames,
so:  (for example (0, 10, 20 ---> 190) is the first scene that has 20 frames and N pedestrians. and then we have 
(200,210,220 ---> 390) is our second scene that has 20 frames and P pedestrians. etc) 
so in the first iteration in the training (framenum = 0), we are gonna be working on the data from frame 0,200,400... 

In other words, nodes_abs for instance, it has the shape (19,N,2), nodes_abs[0] contains the data of the pedestrians from different frames (0,200,400)

**Comment by cunjunyu** [comment]
Yes, that is right.

Thanks.

**Comment by MazenHamdi** [comment]
That's perfectly clear, much appreciated, 
one last question and I close the issue if I may,
how is the number of scenes that form a batch is determined ?

**Comment by cunjunyu** [comment]
> how is the number of scenes that form a batch is determined?
It depends on the number of pedestrians you want in a batch. For example, if you want N pedestrians in a batch, then scenes will be sampled in order to make the number of pedestrians in the range (N, 2N).

**Comment by MazenHamdi** [comment]
Hello again, I was running  through the code, and something caught my attention,
why does the spatial encoder takes as in input, spatial_input_embedded with the shape (1 , N, 32) for each iteration, while the temporal takes as input temporal_input_embedded with shape (N, framenum, 32) ? (framenum in range (seq_length -1) )

**Comment by MazenHamdi** [comment]
Hello again, I was running  through the code, and something caught my attention,
why does the spatial encoder takes as in input, spatial_input_embedded with the shape (N,1, 32) for each iteration, while the temporal takes as input temporal_input_embedded with shape (framenum,1, 32) ? (framenum in range (seqlength -1).
correct me if I am wrong: 
for the spatial encoder, the seq length is N, and the batch size is 1, so basically it's going to map every single pedestrian on a different (query,keys,values), hence we get the social interaction.
for the temporal encoder, the seq length is "framenum" (the number of the frame of the current iteration), and the seq length is N. I don't really see how this works as a temporal independacy per individual ,
thanks in advance

**Comment by cunjunyu** [comment]
> for the spatial encoder, the seq length is N, and the batch size is 1, so basically it's going to map every single pedestrian on a different (query,keys,values), hence we get the social interaction.

You are right about this.

> for the temporal encoder, the seq length is "framenum" (the number of the frame of the current iteration), and the seq length is N. I don't really see how this works as a temporal independacy per individual 

In the spatial transformer, it calculates the attention between different pedestrians in the same frame. In the temporal transformer,  it calculates the attention for the same pedestrian between different frames.
